File: WonderNestApp/lib/models/content_pack.dart
import 'package:json_annotation/json_annotation.dart';

part 'content_pack.g.dart';

@JsonSerializable()
class ContentPackCategory {
  final String id;
  final String name;
  final String? description;
  final int displayOrder;
  final String? iconUrl;
  final String? colorHex;
  final bool isActive;
  final int ageMin;
  final int ageMax;
  final DateTime createdAt;
  final DateTime updatedAt;

  ContentPackCategory({
    required this.id,
    required this.name,
    this.description,
    this.displayOrder = 0,
    this.iconUrl,
    this.colorHex,
    this.isActive = true,
    this.ageMin = 3,
    this.ageMax = 12,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ContentPackCategory.fromJson(Map<String, dynamic> json) => 
      _$ContentPackCategoryFromJson(json);
  Map<String, dynamic> toJson() => _$ContentPackCategoryToJson(this);
}

@JsonSerializable()
class ContentPack {
  final String id;
  final String name;
  final String? description;
  final String? shortDescription;
  final String packType;
  final String? categoryId;
  final ContentPackCategory? category;
  
  // Pricing and availability
  final int priceCents;
  final bool isFree;
  final bool isFeatured;
  final bool isPremium;
  
  // Age and educational info
  final int ageMin;
  final int ageMax;
  final List<String> educationalGoals;
  final List<String> curriculumTags;
  
  // Visual and metadata
  final String? thumbnailUrl;
  final List<String> previewUrls;
  final String? bannerImageUrl;
  final Map<String, String>? colorPalette;
  final String? artStyle;
  final List<String> moodTags;
  
  // Technical metadata
  final int totalAssets;
  final int fileSizeBytes;
  final List<String> supportedPlatforms;
  final String? minAppVersion;
  final String performanceTier;
  
  // Status and timestamps
  final String status;
  final DateTime? publishedAt;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? createdBy;
  
  // Search and discovery
  final String? searchKeywords;
  final double popularityScore;
  final int downloadCount;
  final double ratingAverage;
  final int ratingCount;
  
  // Related data
  final List<ContentPackAsset> assets;
  final UserPackOwnership? userOwnership;

  ContentPack({
    required this.id,
    required this.name,
    this.description,
    this.shortDescription,
    required this.packType,
    this.categoryId,
    this.category,
    this.priceCents = 0,
    this.isFree = false,
    this.isFeatured = false,
    this.isPremium = false,
    this.ageMin = 3,
    this.ageMax = 12,
    this.educationalGoals = const [],
    this.curriculumTags = const [],
    this.thumbnailUrl,
    this.previewUrls = const [],
    this.bannerImageUrl,
    this.colorPalette,
    this.artStyle,
    this.moodTags = const [],
    this.totalAssets = 0,
    this.fileSizeBytes = 0,
    this.supportedPlatforms = const ['ios', 'android', 'web'],
    this.minAppVersion,
    this.performanceTier = 'standard',
    this.status = 'draft',
    this.publishedAt,
    required this.createdAt,
    required this.updatedAt,
    this.createdBy,
    this.searchKeywords,
    this.popularityScore = 0.0,
    this.downloadCount = 0,
    this.ratingAverage = 0.0,
    this.ratingCount = 0,
    this.assets = const [],
    this.userOwnership,
  });

  factory ContentPack.fromJson(Map<String, dynamic> json) => 
      _$ContentPackFromJson(json);
  Map<String, dynamic> toJson() => _$ContentPackToJson(this);

  // Helper methods
  bool get isOwned => userOwnership != null;
  bool get isDownloaded => userOwnership?.downloadStatus == 'completed';
  String get priceDisplay => isFree ? 'Free' : '\$${(priceCents / 100).toStringAsFixed(2)}';
  String get ageRangeDisplay => '$ageMin-$ageMax years';
  String get fileSizeDisplay {
    if (fileSizeBytes < 1024) return '${fileSizeBytes}B';
    if (fileSizeBytes < 1024 * 1024) return '${(fileSizeBytes / 1024).toStringAsFixed(1)}KB';
    return '${(fileSizeBytes / (1024 * 1024)).toStringAsFixed(1)}MB';
  }
}

@JsonSerializable()
class ContentPackAsset {
  final String id;
  final String packId;
  final String name;
  final String? description;
  final String assetType;
  final String fileUrl;
  final String? thumbnailUrl;
  final String? fileFormat;
  final int? fileSizeBytes;
  final int? dimensionsWidth;
  final int? dimensionsHeight;
  final double? durationSeconds;
  final int? frameRate;
  final List<String> tags;
  final Map<String, String>? colorPalette;
  final bool transparencySupport;
  final Map<String, int>? loopPoints;
  final Map<String, dynamic>? interactionConfig;
  final List<String> animationTriggers;
  final int displayOrder;
  final String? groupName;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  ContentPackAsset({
    required this.id,
    required this.packId,
    required this.name,
    this.description,
    required this.assetType,
    required this.fileUrl,
    this.thumbnailUrl,
    this.fileFormat,
    this.fileSizeBytes,
    this.dimensionsWidth,
    this.dimensionsHeight,
    this.durationSeconds,
    this.frameRate,
    this.tags = const [],
    this.colorPalette,
    this.transparencySupport = false,
    this.loopPoints,
    this.interactionConfig,
    this.animationTriggers = const [],
    this.displayOrder = 0,
    this.groupName,
    this.isActive = true,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ContentPackAsset.fromJson(Map<String, dynamic> json) => 
      _$ContentPackAssetFromJson(json);
  Map<String, dynamic> toJson() => _$ContentPackAssetToJson(this);
}

@JsonSerializable()
class UserPackOwnership {
  final String id;
  final String userId;
  final String packId;
  final String? childId;
  final DateTime acquiredAt;
  final String acquisitionType;
  final int purchasePriceCents;
  final String? transactionId;
  final String downloadStatus;
  final int downloadProgress;
  final DateTime? downloadedAt;
  final DateTime? lastUsedAt;
  final int usageCount;
  final bool isFavorite;
  final bool isHidden;
  final List<String> customTags;

  UserPackOwnership({
    required this.id,
    required this.userId,
    required this.packId,
    this.childId,
    required this.acquiredAt,
    this.acquisitionType = 'purchase',
    this.purchasePriceCents = 0,
    this.transactionId,
    this.downloadStatus = 'pending',
    this.downloadProgress = 0,
    this.downloadedAt,
    this.lastUsedAt,
    this.usageCount = 0,
    this.isFavorite = false,
    this.isHidden = false,
    this.customTags = const [],
  });

  factory UserPackOwnership.fromJson(Map<String, dynamic> json) => 
      _$UserPackOwnershipFromJson(json);
  Map<String, dynamic> toJson() => _$UserPackOwnershipToJson(this);
}

@JsonSerializable()
class ContentPackSearchRequest {
  final String? query;
  final String? category;
  final String? packType;
  final int? ageMin;
  final int? ageMax;
  final int? priceMin;
  final int? priceMax;
  final bool? isFree;
  final List<String> educationalGoals;
  final String sortBy;
  final String sortOrder;
  final int page;
  final int size;

  ContentPackSearchRequest({
    this.query,
    this.category,
    this.packType,
    this.ageMin,
    this.ageMax,
    this.priceMin,
    this.priceMax,
    this.isFree,
    this.educationalGoals = const [],
    this.sortBy = 'popularity',
    this.sortOrder = 'desc',
    this.page = 0,
    this.size = 20,
  });

  factory ContentPackSearchRequest.fromJson(Map<String, dynamic> json) => 
      _$ContentPackSearchRequestFromJson(json);
  Map<String, dynamic> toJson() => _$ContentPackSearchRequestToJson(this);
}

@JsonSerializable()
class ContentPackSearchResponse {
  final List<ContentPack> packs;
  final int total;
  final int page;
  final int size;
  final bool hasNext;

  ContentPackSearchResponse({
    required this.packs,
    required this.total,
    required this.page,
    required this.size,
    required this.hasNext,
  });

  factory ContentPackSearchResponse.fromJson(Map<String, dynamic> json) => 
      _$ContentPackSearchResponseFromJson(json);
  Map<String, dynamic> toJson() => _$ContentPackSearchResponseToJson(this);
}

// Pack type enums for type safety
enum ContentPackType {
  characterBundle,
  backdropCollection,
  stickerPack,
  soundEffects,
  musicCollection,
  voicePack,
  emojiPack,
  spriteSheet,
  interactiveObjects,
  particleEffects,
  texturePack,
  animationBundle,
  educationalTheme,
}

enum MediaType {
  imageStatic,
  imageAnimated,
  spriteSheet,
  vectorAnimation,
  audioSound,
  audioMusic,
  audioVoice,
  videoShort,
  interactiveObject,
  particleSystem,
  texture3D,
  model3D,
  fontCustom,
}

extension ContentPackTypeExtension on ContentPackType {
  String get displayName {
    switch (this) {
      case ContentPackType.characterBundle:
        return 'Character Bundle';
      case ContentPackType.backdropCollection:
        return 'Backdrop Collection';
      case ContentPackType.stickerPack:
        return 'Sticker Pack';
      case ContentPackType.soundEffects:
        return 'Sound Effects';
      case ContentPackType.musicCollection:
        return 'Music Collection';
      case ContentPackType.voicePack:
        return 'Voice Pack';
      case ContentPackType.emojiPack:
        return 'Emoji Pack';
      case ContentPackType.spriteSheet:
        return 'Sprite Sheet';
      case ContentPackType.interactiveObjects:
        return 'Interactive Objects';
      case ContentPackType.particleEffects:
        return 'Particle Effects';
      case ContentPackType.texturePack:
        return 'Texture Pack';
      case ContentPackType.animationBundle:
        return 'Animation Bundle';
      case ContentPackType.educationalTheme:
        return 'Educational Theme';
    }
  }
}

File: Wonder Nest Backend/src/main/kotlin/com/wondernest/models/ContentPack.kt
package com.wondernest.models

import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import java.math.BigDecimal
import java.time.Instant
import java.util.UUID

@Serializable
data class ContentPackCategory(
    @Contextual val id: UUID,
    val name: String,
    val description: String? = null,
    val displayOrder: Int = 0,
    val iconUrl: String? = null,
    val colorHex: String? = null,
    val isActive: Boolean = true,
    val ageMin: Int = 3,
    val ageMax: Int = 12,
    @Contextual val createdAt: Instant,
    @Contextual val updatedAt: Instant
)

@Serializable
data class ContentPack(
    @Contextual val id: UUID,
    val name: String,
    val description: String? = null,
    val shortDescription: String? = null,
    val packType: String,
    @Contextual val categoryId: UUID? = null,
    val category: ContentPackCategory? = null,
    
    // Pricing and availability
    val priceCents: Int = 0,
    val isFree: Boolean = false,
    val isFeatured: Boolean = false,
    val isPremium: Boolean = false,
    
    // Age and educational info
    val ageMin: Int = 3,
    val ageMax: Int = 12,
    val educationalGoals: List<String> = emptyList(),
    val curriculumTags: List<String> = emptyList(),
    
    // Visual and metadata
    val thumbnailUrl: String? = null,
    val previewUrls: List<String> = emptyList(),
    val bannerImageUrl: String? = null,
    val colorPalette: Map<String, String>? = null,
    val artStyle: String? = null,
    val moodTags: List<String> = emptyList(),
    
    // Technical metadata
    val totalAssets: Int = 0,
    val fileSizeBytes: Long = 0,
    val supportedPlatforms: List<String> = listOf("ios", "android", "web"),
    val minAppVersion: String? = null,
    val performanceTier: String = "standard",
    
    // Status and timestamps
    val status: String = "draft",
    @Contextual val publishedAt: Instant? = null,
    @Contextual val createdAt: Instant,
    @Contextual val updatedAt: Instant,
    @Contextual val createdBy: UUID? = null,
    
    // Search and discovery
    val searchKeywords: String? = null,
    @Contextual val popularityScore: BigDecimal = BigDecimal("0.00"),
    val downloadCount: Long = 0,
    @Contextual val ratingAverage: BigDecimal = BigDecimal("0.00"),
    val ratingCount: Int = 0,
    
    // Related data
    val assets: List<ContentPackAsset> = emptyList(),
    val userOwnership: UserPackOwnership? = null,
    val userReview: ContentPackReview? = null
)

@Serializable
data class ContentPackAsset(
    @Contextual val id: UUID,
    @Contextual val packId: UUID,
    
    // Basic asset info
    val name: String,
    val description: String? = null,
    val assetType: String,
    val fileUrl: String,
    val thumbnailUrl: String? = null,
    
    // File technical details
    val fileFormat: String? = null,
    val fileSizeBytes: Int? = null,
    val dimensionsWidth: Int? = null,
    val dimensionsHeight: Int? = null,
    @Contextual val durationSeconds: BigDecimal? = null,
    val frameRate: Int? = null,
    
    // Creative metadata
    val tags: List<String> = emptyList(),
    val colorPalette: Map<String, String>? = null,
    val transparencySupport: Boolean = false,
    val loopPoints: Map<String, Int>? = null,
    
    // Interactive properties
    val interactionConfig: Map<String, String>? = null,
    val animationTriggers: List<String> = emptyList(),
    
    // Ordering and grouping
    val displayOrder: Int = 0,
    val groupName: String? = null,
    
    // Status
    val isActive: Boolean = true,
    @Contextual val createdAt: Instant,
    @Contextual val updatedAt: Instant
)

@Serializable
data class UserPackOwnership(
    @Contextual val id: UUID,
    @Contextual val userId: UUID,
    @Contextual val packId: UUID,
    @Contextual val childId: UUID? = null,
    
    // Purchase/acquisition info
    @Contextual val acquiredAt: Instant,
    val acquisitionType: String = "purchase",
    val purchasePriceCents: Int = 0,
    val transactionId: String? = null,
    
    // Download and usage
    val downloadStatus: String = "pending",
    val downloadProgress: Int = 0,
    @Contextual val downloadedAt: Instant? = null,
    @Contextual val lastUsedAt: Instant? = null,
    val usageCount: Int = 0,
    
    // Preferences
    val isFavorite: Boolean = false,
    val isHidden: Boolean = false,
    val customTags: List<String> = emptyList()
)

@Serializable
data class ContentPackUsage(
    @Contextual val id: UUID,
    @Contextual val userId: UUID,
    @Contextual val childId: UUID? = null,
    @Contextual val packId: UUID,
    @Contextual val assetId: UUID? = null,
    
    // Usage context
    val usedInFeature: String? = null,
    @Contextual val sessionId: UUID? = null,
    val usageDurationSeconds: Int? = null,
    
    // Metadata
    @Contextual val usedAt: Instant,
    val usageMetadata: Map<String, String>? = null
)

@Serializable
data class ContentPackCollection(
    @Contextual val id: UUID,
    val name: String,
    val description: String? = null,
    val collectionType: String = "bundle",
    
    // Pricing and availability
    val priceCents: Int = 0,
    val discountPercentage: Int = 0,
    
    // Visual
    val thumbnailUrl: String? = null,
    val bannerImageUrl: String? = null,
    
    // Status
    val isActive: Boolean = true,
    val isFeatured: Boolean = false,
    @Contextual val availableFrom: Instant? = null,
    @Contextual val availableUntil: Instant? = null,
    
    @Contextual val createdAt: Instant,
    @Contextual val updatedAt: Instant,
    
    // Related data
    val packs: List<ContentPack> = emptyList()
)

@Serializable
data class ContentPackReview(
    @Contextual val id: UUID,
    @Contextual val packId: UUID,
    @Contextual val userId: UUID,
    
    // Review content
    val rating: Int,
    val reviewText: String? = null,
    val reviewTitle: String? = null,
    
    // Helpful metrics
    val helpfulCount: Int = 0,
    val notHelpfulCount: Int = 0,
    
    // Child context (anonymous)
    val childAgeRange: String? = null,
    val usedFeatures: List<String> = emptyList(),
    
    // Moderation
    val isApproved: Boolean = false,
    val isFeatured: Boolean = false,
    @Contextual val moderatedAt: Instant? = null,
    @Contextual val moderatedBy: UUID? = null,
    
    @Contextual val createdAt: Instant,
    @Contextual val updatedAt: Instant
)

// Request/Response DTOs
@Serializable
data class ContentPackSearchRequest(
    val query: String? = null,
    val category: String? = null,
    val packType: String? = null,
    val ageMin: Int? = null,
    val ageMax: Int? = null,
    val priceMin: Int? = null,
    val priceMax: Int? = null,
    val isFree: Boolean? = null,
    val educationalGoals: List<String> = emptyList(),
    val sortBy: String = "popularity",
    val sortOrder: String = "desc",
    val page: Int = 0,
    val size: Int = 20
)

@Serializable
data class ContentPackSearchResponse(
    val packs: List<ContentPack>,
    val total: Long,
    val page: Int,
    val size: Int,
    val hasNext: Boolean
)

@Serializable
data class PackPurchaseRequest(
    @Contextual val packId: UUID,
    @Contextual val childId: UUID? = null,
    val paymentMethod: String? = null
)

@Serializable
data class PackPurchaseResponse(
    val success: Boolean,
    val transactionId: String? = null,
    val ownership: UserPackOwnership? = null,
    val error: String? = null
)